File: src/Offering/Offering/Controllers/OfferController.cs
using Microsoft.AspNetCore.Mvc;
using Offering.Dtos;
using Offering.Models;
using Offering.Repositories;

namespace Offering.Controllers;

[ApiController]
[Route("/api/[controller]")]
[Produces("application/json")]
public class OfferController : Controller
{
    private readonly IOfferRepository _offerRepository;

    private readonly IProductRepository _productRepository;

    public OfferController(IOfferRepository offerRepository, IProductRepository productRepository)
    {
        _offerRepository = offerRepository;
        _productRepository = productRepository;
    }

    [HttpGet]
    public async Task<IActionResult> ListOffers()
    {
        var offerResult = await _offerRepository.FindAll();
        return Ok(offerResult);
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> GetOffer(Guid id)
    {
        var offerResult = await _offerRepository.FindById(id);
        return Ok(offerResult);
    }


    [HttpPost("single")]
    public async Task<IActionResult> CreateSingleOffer([FromBody] CreateSingleOfferRequestDto createSingleOfferRequestDto)
    {
        var price = Price.CreateFromGross(createSingleOfferRequestDto.grossPrice, createSingleOfferRequestDto.taxValue,
            createSingleOfferRequestDto.currency);
        var product = await _productRepository.FindById(createSingleOfferRequestDto.productId);
        var offer = SingleOffer.Create(createSingleOfferRequestDto.name, price, createSingleOfferRequestDto.startDate,
            createSingleOfferRequestDto.endDate, product);

        var createdOffer = await _offerRepository.Add(offer);
        return Ok(createdOffer);
    }
    
    [HttpPost("package")]
    public async Task<IActionResult> CreatePackageOffer([FromBody] CreatePackageOfferRequestDto createPackageOfferRequestDto)
    {
        var price = Price.CreateFromGross(createPackageOfferRequestDto.grossPrice, createPackageOfferRequestDto.taxValue,
            createPackageOfferRequestDto.currency);

        var products = new List<Product>();
        foreach (var productId in createPackageOfferRequestDto.productIds)
        {
            var product = await _productRepository.FindById(productId);
            products.Add(product);
        }
        var offer = PackageOffer.Create(createPackageOfferRequestDto.name, price, createPackageOfferRequestDto.startDate,
            createPackageOfferRequestDto.endDate, products);

        var createdOffer = await _offerRepository.Add(offer);
        return Ok(createdOffer);
    }
    
    [HttpDelete("{id}")]
    public async Task<IActionResult> DeleteOffer(Guid id)
    {
        var offer = await _offerRepository.FindById(id);
        await _offerRepository.Delete(offer);
        return Ok();
    }
}

File: src/OfferSB/src/main/java/com/example/offersb/controllers/OfferController.java
package com.example.offersb.controllers;

import com.example.offersb.Dtos.CreatePackageOfferRequestDto;
import com.example.offersb.Dtos.CreateSingleOfferRequestDto;
import com.example.offersb.models.*;
import com.example.offersb.repositories.ILocalizationRepository;
import com.example.offersb.repositories.IOfferRepository;
import com.example.offersb.repositories.IProductRepository;
import org.springframework.web.bind.annotation.*;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/offers")
public class OfferController {

    private final IOfferRepository offerRepository;

    private final IProductRepository productRepository;

    private final ILocalizationRepository localizationRepository;

    public OfferController(IOfferRepository offerRepository, IProductRepository productRepository, ILocalizationRepository localizationRepository) {
        this.offerRepository = offerRepository;
        this.productRepository = productRepository;
        this.localizationRepository = localizationRepository;
    }

    @GetMapping()
    public List<Offer> getAllOffers() {
        return offerRepository.findAll();
    }

    @GetMapping("/{id}")
    public Offer getOfferById(@PathVariable("id") UUID id) {
        return offerRepository.findById(id).orElse(null);
    }

    @PostMapping("single")
    public SingleOffer createSingleOffer(@RequestBody CreateSingleOfferRequestDto offerDto) {
        var price = Price.CreateFromGross(offerDto.grossPrice(), offerDto.taxValue());
        var product = productRepository.findById(offerDto.productId()).orElse(null);
        var localization = localizationRepository.findById("DE");
        var offer = SingleOffer.create(UUID.randomUUID(), offerDto.name(), price, offerDto.startDate(), offerDto.endDate(), product, localization.get());
        return offerRepository.save(offer);
    }

    @PostMapping("package")
    public PackageOffer createPackageOffer(@RequestBody CreatePackageOfferRequestDto offerDto) {
        var price = Price.CreateFromGross(offerDto.grossPrice(), offerDto.taxValue());

        var products = new ArrayList<Product>();
        for (var productId : offerDto.productIds()) {
            var product = productRepository.findById(productId).orElse(null);
            products.add(product);
        }
        var localization = localizationRepository.findById("DE");
        var offer = PackageOffer.create(UUID.randomUUID(), offerDto.name(), price, offerDto.startDate(), offerDto.endDate(), products, localization.get());
        return offerRepository.save(offer);
    }

    @DeleteMapping("/{id}")
    public void deleteOffer(@PathVariable("id") UUID id) {
        offerRepository.deleteById(id);
    }
}
